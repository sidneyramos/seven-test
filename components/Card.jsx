import React, { Fragment, useEffect } from "react";
import styled from "styled-components";
import tileImage from "../static/tile.jpg";
import logoImage from "../static/logo.png";
import { Textfit } from "react-textfit";
import fitty from "fitty";

const classNames = require("classnames");
require("typeface-montserrat");

const CardContainer = styled.div`
  height: 0;
  overflow: hidden;
  padding-top: ${props =>
    props.width && props.height
      ? `${(props.height / props.width) * 100}%`
      : "100%"};
  background: #f0f0f0;
  background-image: url("${props => props.backgroundImage}");
  background-size: cover;
  position: relative;
  font-family: Montserrat;

  &.hover {
    cursor: pointer;
    
    .overlay-container {
      opacity: 0;
      transition: all 0.3s;
    }

    &:hover {
      .overlay-container {
        opacity: 1;
      }
    }
  }
`;

const InnerCardContainer = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
`;

const FlexBox = styled.div`
  display: flex;
  height: 100%;
  flex-direction: column;
  justify-content: flex-end;
`;

const OverlayContainer = styled.div`
  display: flex;
  align-items: center;
  height: calc(23%);
  background-color: rgba(0, 0, 0, 0.5);
  padding: 2%;
  color: white;
`;

const OverlayTitle = styled.h1`
  margin: 0;
  line-height: 1.2;
`;

const OverlayTitleContainer = styled.div`
  height: 50%;
  width: 50%;
`;

const LogoImage = styled.img`
  height: 100%;
  width: auto;
`;

const Card = ({
  children,
  image = "http://via.placeholder.com/640x360",
  logo,
  title,
  isHover,
  ...props
}) => {
  let backgroundImage = new Image();
  backgroundImage.src = image;
  useEffect(() => {
    fitty(".fit");
  });

  return (
    <CardContainer
      className={classNames([
        "card-container",
        {
          hover: isHover
        }
      ])}
      width={backgroundImage.width}
      height={backgroundImage.height}
      backgroundImage={image}
      {...props}
    >
      <InnerCardContainer className="inner-card-container">
        <FlexBox>
          {title && (
            <OverlayContainer className="overlay-container">
              {logo && <LogoImage src={logo.src} alt={logo.alt} />}
              <OverlayTitleContainer>
                <OverlayTitle className="fit">{title}</OverlayTitle>
              </OverlayTitleContainer>
            </OverlayContainer>
          )}
        </FlexBox>
      </InnerCardContainer>
    </CardContainer>
  );
};

export default Card;
