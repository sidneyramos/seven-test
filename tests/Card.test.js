import React from "react";
import renderer from "react-test-renderer";
import Card from "../components/Card";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Card
        image="https://sidneyramos.github.io/seven-test/static/media/tile.25b4e634.jpg"
        logo={{
          src:
            "https://sidneyramos.github.io/seven-test/static/media/logo.2a3f011a.png",
          alt: "Seven West Media"
        }}
        title="Home and Away"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly without props", () => {
  const tree = renderer.create(<Card />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly without image", () => {
  const tree = renderer
    .create(
      <Card
        logo={{
          src:
            "https://sidneyramos.github.io/seven-test/static/media/logo.2a3f011a.png",
          alt: "Seven West Media"
        }}
        title="Home and Away"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly without logo", () => {
  const tree = renderer
    .create(
      <Card
        image="https://sidneyramos.github.io/seven-test/static/media/tile.25b4e634.jpg"
        title="Home and Away"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly without title", () => {
  const tree = renderer
    .create(
      <Card
        image="https://sidneyramos.github.io/seven-test/static/media/tile.25b4e634.jpg"
        logo={{
          src:
            "https://sidneyramos.github.io/seven-test/static/media/logo.2a3f011a.png",
          alt: "Seven West Media"
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with hover props", () => {
  const tree = renderer
    .create(
      <Card
        image="https://sidneyramos.github.io/seven-test/static/media/tile.25b4e634.jpg"
        logo={{
          src:
            "https://sidneyramos.github.io/seven-test/static/media/logo.2a3f011a.png",
          alt: "Seven West Media"
        }}
        title="Home and Away"
        isHover
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
