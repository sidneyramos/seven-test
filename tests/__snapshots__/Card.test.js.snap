// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-AxjAm imRVdv card-container"
  height={0}
  width={0}
>
  <div
    className="sc-AxirZ qSyhk inner-card-container"
  >
    <div
      className="sc-AxiKw isPqKo"
    >
      <div
        className="sc-AxhCb eDRkHZ overlay-container"
      >
        <img
          alt="Seven West Media"
          className="sc-AxheI lddWen"
          src="https://sidneyramos.github.io/seven-test/static/media/logo.2a3f011a.png"
        />
        <div
          className="sc-AxgMl bsNrIm"
        >
          <h1
            className="sc-AxhUy ivyjjC fit"
          >
            Home and Away
          </h1>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with hover props 1`] = `
<div
  className="sc-AxjAm imRVdv card-container hover"
  height={0}
  width={0}
>
  <div
    className="sc-AxirZ qSyhk inner-card-container"
  >
    <div
      className="sc-AxiKw isPqKo"
    >
      <div
        className="sc-AxhCb eDRkHZ overlay-container"
      >
        <img
          alt="Seven West Media"
          className="sc-AxheI lddWen"
          src="https://sidneyramos.github.io/seven-test/static/media/logo.2a3f011a.png"
        />
        <div
          className="sc-AxgMl bsNrIm"
        >
          <h1
            className="sc-AxhUy ivyjjC fit"
          >
            Home and Away
          </h1>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly without image 1`] = `
<div
  className="sc-AxjAm UnUtx card-container"
  height={0}
  width={0}
>
  <div
    className="sc-AxirZ qSyhk inner-card-container"
  >
    <div
      className="sc-AxiKw isPqKo"
    >
      <div
        className="sc-AxhCb eDRkHZ overlay-container"
      >
        <img
          alt="Seven West Media"
          className="sc-AxheI lddWen"
          src="https://sidneyramos.github.io/seven-test/static/media/logo.2a3f011a.png"
        />
        <div
          className="sc-AxgMl bsNrIm"
        >
          <h1
            className="sc-AxhUy ivyjjC fit"
          >
            Home and Away
          </h1>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly without logo 1`] = `
<div
  className="sc-AxjAm imRVdv card-container"
  height={0}
  width={0}
>
  <div
    className="sc-AxirZ qSyhk inner-card-container"
  >
    <div
      className="sc-AxiKw isPqKo"
    >
      <div
        className="sc-AxhCb eDRkHZ overlay-container"
      >
        <div
          className="sc-AxgMl bsNrIm"
        >
          <h1
            className="sc-AxhUy ivyjjC fit"
          >
            Home and Away
          </h1>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly without props 1`] = `
<div
  className="sc-AxjAm UnUtx card-container"
  height={0}
  width={0}
>
  <div
    className="sc-AxirZ qSyhk inner-card-container"
  >
    <div
      className="sc-AxiKw isPqKo"
    />
  </div>
</div>
`;

exports[`renders correctly without title 1`] = `
<div
  className="sc-AxjAm imRVdv card-container"
  height={0}
  width={0}
>
  <div
    className="sc-AxirZ qSyhk inner-card-container"
  >
    <div
      className="sc-AxiKw isPqKo"
    />
  </div>
</div>
`;
